Add ObjectTagging Support

nitisht opened the pull request:
## Description
This PR adds support for AWS S3 ObjectTagging API as explained here https://docs.aws.amazon.com/AmazonS3/latest/dev/object-tagging.html

## Motivation and Context
AWS S3 Compatibility

## How to test this PR?
Use any S3 compatible system to send object tagging requests.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -2674,3 +2721,112 @@ func (api objectAPIHandlers) GetObjectRetentionHandler(w http.ResponseWriter, r
+
+	tagging, err := tagging.ParseTagging(io.LimitReader(r.Body, r.ContentLength))
+	if err != nil {
+		writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrMalformedXML), r.URL, guessIsBrowserReq(r))
```
Look at how we now send internal parsing errors back to users just like policy.ParseConfig()

nitisht replied:
done

harshavardhana commented on cmd/object-api-utils.go:
```diff
@@ -239,10 +239,35 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
 
+// removeTags deletes all tag entries (that start with x-amz-tag-)
+// from the given map and returns the same map
+func removeTags(metadata map[string]string) map[string]string {
```
Always make a copy of map to avoid races

nitisht replied:
done

Praveenrajmani commented on cmd/handler-utils.go:
```diff
@@ -174,6 +186,37 @@ func extractMetadataFromMap(ctx context.Context, v map[string][]string, m map[st
+		for key := range tags {
+			if value, ok := tags[key]; ok {
+				// Add the "X-Amz-Tag" prefix
+				parsedTags[xhttp.AmzObjectTagKeyPrefix+key] = strings.Join(value, ",")
```
should we have the check for the prefix already? Like, adding the prefix only if it is not found already.. like we do for user-defined metadata? @nitisht

Praveenrajmani replied:
Never mind - Clients should have taken care of these things

Praveenrajmani commented on cmd/handler-utils.go:
```diff
@@ -120,6 +121,17 @@ var userMetadataKeyPrefixes = []string{
 	"X-Minio-Meta-",
 }
 
+func filterTagKeyPrefix(ctx context.Context, meta map[string]string) map[string]string {
```
looks like we are not using this function anywhere else, we can remove this

Praveenrajmani reviewed (approved):
Tested and LGTM otherwise

kannappanr reviewed (changes requested):
Please fix the issues listed in the comments

klauspost commented on cmd/gateway-unsupported.go:
```diff
@@ -173,6 +174,24 @@ func (a GatewayUnsupported) GetMetrics(ctx context.Context) (*Metrics, error) {
 	return &Metrics{}, NotImplemented{}
 }
 
+// PutObjectTag - no op
```
```suggestion
// PutObjectTag - not implemented.
```

klauspost commented on cmd/gateway-unsupported.go:
```diff
@@ -173,6 +174,24 @@ func (a GatewayUnsupported) GetMetrics(ctx context.Context) (*Metrics, error) {
+	return NotImplemented{}
+}
+
+// GetObjectTag - no op
```
```suggestion
// GetObjectTag - not implemented.
```

klauspost commented on cmd/gateway-unsupported.go:
```diff
@@ -173,6 +174,24 @@ func (a GatewayUnsupported) GetMetrics(ctx context.Context) (*Metrics, error) {
+	return tagging.Tagging{}, NotImplemented{}
+}
+
+// DeleteObjectTag - no op
```
```suggestion
// DeleteObjectTag - not implemented.
```

klauspost commented on cmd/handler-utils.go:
```diff
@@ -70,29 +71,29 @@ var supportedHeaders = []string{
-func isMetadataCopy(h http.Header) bool {
-	return h.Get(xhttp.AmzMetadataDirective) == "COPY"
+// Check if the directive COPY is requested.
+func isDirectiveCopy(value string) bool {
```
Do we really need this? It literally takes more characters to write `isDirectiveCopy(v)` than `v == "COPY"`.

I would add constants for the types if we don't have them.

If you want to keep this, at least it could be `value == "COPY" || value == ""` since you write it is the same above. This is of course a functional change.

nitisht replied:
I sort of liked the funcs, but changed it to `value == "COPY" || value == ""`

klauspost commented on cmd/handler-utils.go:
```diff
@@ -70,29 +71,29 @@ var supportedHeaders = []string{
+func isDirectiveValid(v string) bool {
+	if v != "" {
+		// Check if set metadata-directive is valid.
+		return (isDirectiveCopy(v) || isDirectiveReplace(v))
```
```suggestion
		return isDirectiveCopy(v) || isDirectiveReplace(v)
```

klauspost commented on cmd/handler-utils.go:
```diff
@@ -174,6 +175,37 @@ func extractMetadataFromMap(ctx context.Context, v map[string][]string, m map[st
+		}
+		parsedTags := make(map[string]string, len(tags))
+		for key := range tags {
+			if value, ok := tags[key]; ok {
```
Why are we doing this check? We are ranging over tags, so why shouldn't a given key exist?

klauspost commented on cmd/object-api-utils.go:
```diff
@@ -239,10 +239,37 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
+// removeTags deletes all tag entries (that start with x-amz-tag-)
+// from the given map and returns the map
+func removeTags(metadata map[string]string) map[string]string {
+	var newTags = make(map[string]string)
```
```suggestion
	var newTags = make(map[string]string, len(metadata))
```

klauspost commented on cmd/object-handlers.go:
```diff
@@ -2674,3 +2721,113 @@ func (api objectAPIHandlers) GetObjectRetentionHandler(w http.ResponseWriter, r
+		return
+	}
+
+	w.(http.Flusher).Flush()
```
Is this needed? Also I don't like unchecked type assertions.

klauspost commented on cmd/xl-v1-object.go:
```diff
@@ -960,3 +961,60 @@ func (xl xlObjects) ListObjectsV2(ctx context.Context, bucket, prefix, continuat
+		return toObjectErr(err, bucket, object)
+	}
+
+	// Rename atomically `xl.json` from tmp location to destination for each disk.
```
```suggestion
	// Atomically rename `xl.json` from tmp location to destination for each disk.
```

klauspost commented on pkg/tagging/tagging.go:
```diff
@@ -0,0 +1,83 @@
+
+const maxTags = 10
+
+// error returned when more than 10 object tags are supplied
```
Placed wrong?

klauspost commented on pkg/tagging/tagging.go:
```diff
@@ -0,0 +1,83 @@
+	"io"
+)
+
+const maxTags = 10
```
Is this an S3 limit? Maybe add reference.

klauspost commented on pkg/tagging/tagging.go:
```diff
@@ -0,0 +1,83 @@
+	return nil
+}
+
+// ParseTagging - parses data in given reader to Tagging interface.
```
A bit more details, please.

klauspost commented on pkg/tagging/tagset.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// ContainsDuplicate - returns true if duplicate keys are present in TagSet
+func (t TagSet) ContainsDuplicate(key string) bool {
```
Simpler and returns positive earlier:
```
// ContainsDuplicate - returns true if the supplied key is present more than once in TagSet.
func (t TagSet) ContainsDuplicate(key string) bool {
	var found bool
	for _, tag := range t.Tags {
		if tag.Key == key {
			if found {
				return true
			}
			found = true
		}
	}
	return false
}
```

klauspost commented on pkg/tagging/tag.go:
```diff
@@ -0,0 +1,64 @@
+// validateValue - checks if value is valid or not.
+func (t Tag) validateValue() error {
+	// cannot be longer than tagValueLengthMax characters
+	if len(string(t.Value)) > tagValueLengthMax {
```
Value is already string.

This counts code length and not characters. What is the limit?

`if len([]rune(t.Value)) > tagValueLengthMax` will sorta count characters.

The same for the other checks.

klauspost commented on cmd/object-api-utils.go:
```diff
@@ -239,10 +239,37 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
+// extractTagsFromMeta creates a new map with all the tag entries
+// (that start with x-amz-tag-) and returns the new map
+func extractTagsFromMeta(metadata map[string]string) map[string]string {
+	tags := make(map[string]string)
```
```suggestion
	tags := make(map[string]string, len(metadata))
```

klauspost commented on cmd/object-api-utils.go:
```diff
@@ -239,10 +239,37 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
+// removeTags deletes all tag entries (that start with x-amz-tag-)
+// from the given map and returns the map
+func removeTags(metadata map[string]string) map[string]string {
+	var metaWithoutTags = make(map[string]string)
```
```suggestion
	var metaWithoutTags = make(map[string]string, len(metadata))
```

nitisht replied:
We'll be over allocating if we do this, just the tags within the metadata map is bound to be fewer in number than the whole metadata. Is that better than init without a size?

klauspost replied:
Yeah, a slight overallocation is better than having to reallocate+copy, possibly multiple times.

It is only the map structure that is allocated, so it is not that big of an allocation.

klauspost commented on cmd/fs-v1.go:
```diff
@@ -1235,6 +1237,62 @@ func (fs *FSObjects) ListObjects(ctx context.Context, bucket, prefix, marker, de
+		return tagging.Tagging{}, err
+	}
+
+	tags := tagging.Tagging{}
```
Prealloc `tags.TagSet.Tag` with needed capacity.

klauspost commented on pkg/tagging/tagging.go:
```diff
@@ -0,0 +1,83 @@
+
+// Tagging - object tagging interface
+type Tagging struct {
+	XMLName xml.Name `xml:"http://s3.amazonaws.com/doc/2006-03-01/ Tagging"`
```
What does the tag indicate? Is this the name of the tag?

kannappanr replied:
Yes

harshavardhana commented on cmd/object-api-interface.go:
```diff
@@ -125,4 +126,9 @@ type ObjectLayer interface {
 	IsReady(ctx context.Context) bool
+
+	// ObjectTagging operations
+	PutObjectTag(context.Context, string, string, map[string]string) error
```
```suggestion
	PutObjectTag(context.Context, string, string, url.Values) error
```

harshavardhana reviewed (changes requested):
As agreed let's remove the references for x-amz-tag prefix instead keep x-amz-tagging as is.

Commit:
Add ObjectTagging Support

This PR adds support for AWS S3 ObjectTagging API as explained here
https://docs.aws.amazon.com/AmazonS3/latest/dev/object-tagging.html

Co-Authored-By: Klaus Post <klauspost@gmail.com>

Fixes #8446

harshavardhana commented on go.mod:
```diff
@@ -103,6 +103,7 @@ require (
 	github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
 	github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2 // indirect
 	go.etcd.io/bbolt v1.3.3 // indirect
+	go.opencensus.io v0.21.0
```
there is no reason to include this as direct dependency @nitisht can you remove this change?

harshavardhana commented on cmd/api-headers.go:
```diff
@@ -95,6 +96,13 @@ func setObjectHeaders(w http.ResponseWriter, objInfo ObjectInfo, rs *HTTPRangeSp
 	}
+
+	// Set tag count if object has tags
+	tagCount := len(strings.Split(objInfo.UserTags, "&"))
```
This may not be a valid way to calculate this, you should parse `objInfo.UserTags` with url.ParseQuery() and then calculate the `len(url.Values)`

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -174,6 +177,45 @@ func extractMetadataFromMap(ctx context.Context, v map[string][]string, m map[st
+	parsedTags := make([]tagging.Tag, len(tags))
+	for k, v := range tags {
+		parsedTags[idx].Key = k
+		parsedTags[idx].Value = strings.Join(v, ",")
```
Each tag has only one value so we don't need to use strings.Join instead just tags.Get(k) would work @nitisht

harshavardhana commented on cmd/object-api-utils.go:
```diff
@@ -239,10 +239,19 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
+// removeTags deletes the tag entry (x-amz-tagging)
+// from the given map and returns the map
+func removeTags(metadata map[string]string) map[string]string {
+	delete(metadata, xhttp.AmzObjectTagging)
```
@nitisht  there is no reason to return metadata its a memory can be done in one line - you don't need this function.

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -969,11 +970,68 @@ func (xl xlObjects) ListObjectsV2(ctx context.Context, bucket, prefix, continuat
+
+	for i, xlMeta := range metaArr {
+		// clean xlMeta.Meta of all tags, before updating the new tags
+		newMeta := removeTags(xlMeta.Meta)
```
No reason to do this, single key update doesn't require removal just update xlMeta.Meta[xhttp.AmzObjectTagging] = tags

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -969,11 +970,68 @@ func (xl xlObjects) ListObjectsV2(ctx context.Context, bucket, prefix, continuat
+		if tags != "" {
+			newMeta[xhttp.AmzObjectTagging] = tags
+		}
+		metaArr[i].Meta = newMeta
```
This shall be just metaArr[i].Meta = xlMeta.Meta

harshavardhana commented on cmd/object-api-utils.go:
```diff
@@ -239,6 +239,8 @@ func getCompleteMultipartMD5(parts []CompletePart) string {
+	// Remove Tags
+	delete(metadata, xhttp.AmzObjectTagging)
 	// Clean meta etag keys 'md5Sum', 'etag', "expires".
 	return cleanMetadataKeys(metadata, "md5Sum", "etag", "expires")
```
Interestingly enough only one function is necessary here 

```
return cleanMetadataKeys(metadata, "md5Sum", "etag", "expires", xhttp.AmzObjectTagging, xhttp.AmzStorageClass)
```

nitisht replied:
:+1:

nitisht replied:
`removeStandardStorageClass` is needed because the filtering is based on the value instead of key

klauspost commented on cmd/handler-utils.go:
```diff
@@ -174,6 +177,45 @@ func extractMetadataFromMap(ctx context.Context, v map[string][]string, m map[st
+	return "", nil
+}
+
+func extractTagsFromString(tagStr string) (tagging.Tagging, error) {
```
Shouldn't we move this to the `tagging` package? Seems like it would belong there.

`func TaggingFromString(tagStr string) (Tagging, error)`

nitisht replied:
this is moved

klauspost reviewed (approved):
Very nice! Just a single thing. Not a blocker.
